internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// idkey supplies an identity key for a machine, container or device. This
+// package provides an implementation based on an Ed25519 based key, currently
+// only supporting a file based key storage solution. This can later be
+// extended to support more secure hardware backed keystores, such as TPM,
+// OP-TEE or YubiKey.
+package idkey
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/base32"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+const identityKeyFile = "key.pem"
+
+// Identity key must implement a crypto signer.
+var _ crypto.Signer = (*IDKey)(nil)
+
+type IDKey struct {
+	keyDir string
+	key    any
+}
+
+// New checks if an existing private identity key exists, and loads the key
+// if it does. It creates a new private identity key and persists it to disk
+// if no key was found. Cases where explicit control is desired on when to
+// generate or load can use GenerateKey and LoadKey directly.
+func New(keyDir string) (*IDKey, error) {
+	keyPath := filepath.Join(keyDir, identityKeyFile)
+	exists, err := pathExists(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access key file %q: %w", keyPath, err)
+	}
+	if exists {
+		return LoadKey(keyDir)
+	}
+	return GenerateKey(keyDir)
+}
+
+// GenerateKey generates a new identity key and persists it to disk. This
+// function should only ever be called on the first boot otherwise the
+// existing identity will be overwritten.
+//
+// This function is equivalent to running:
+//
+//	openssl genpkey -algorithm Ed25519 -out key.pem
+func GenerateKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	err := key.createDir()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create identity directory: %w", err)
+	}
+	err = key.newEd25519()
```

Comment:
This method has the side effect of setting `key.key`, which given the name, seems odd ("new" functions usually create a new object and return it). Plus, the method is only called once and basically wraps a single function call. What about inlining it here:

```go
_, key.key, err = ed25519.GenerateKey(rand.Reader)
```

Reply:
Fixed. Thanks. Resolving.